p2p/libp2p/factory/peerDiscovererFactory.go

```diff
@@ -0,0 +1,58 @@
+package factory
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/p2p/libp2p/discovery"
+)
+
+type peerDiscovererFactory struct {
+	p2pConfig config.P2PConfig
+}
+
+// NewPeerDiscovererFactory creates a new instance of peer discovery factory
+func NewPeerDiscovererFactory(pConfig config.P2PConfig) *peerDiscovererFactory {
+	return &peerDiscovererFactory{
+		p2pConfig: pConfig,
+	}
+}
+
+// CreatePeerDiscoverer generates an implementation of PeerDiscoverer by parsing the p2pConfig struct
+// Errors if config is badly formatted
+func (pdf *peerDiscovererFactory) CreatePeerDiscoverer() (p2p.PeerDiscoverer, error) {
+	if pdf.p2pConfig.KadDhtPeerDiscovery.Enabled {
+		return pdf.createKadDhtPeerDiscoverer()
+	}
+
+	return discovery.NewNullDiscoverer(), nil
+}
+
+func (pdf *peerDiscovererFactory) createKadDhtPeerDiscoverer() (p2p.PeerDiscoverer, error) {
+	if pdf.p2pConfig.KadDhtPeerDiscovery.RefreshIntervalInSec < 1 {
+		return nil, fmt.Errorf("%w for RefreshIntervalInSec, expected value > 0", p2p.ErrInvalidValue)
+	}
+	if pdf.p2pConfig.KadDhtPeerDiscovery.RoutingTableRefreshIntervalInSec < 1 {
```

Comment:
I don't like that you check 2 times the same config values. Can you skip the checking in one of them ?
Right now, you check the config and then send them to the NewKadDhtPeerDiscoverer after converting to time.Duration so right now there is no way that the values would arrive broken there

Reply:
Agree

Reply:
+1 check in a single time.

Reply:
There are some subtleties regarding factories. The pattern should be uniformized. Although it works in this use case (create method is called just after the factory has been instantiated, it might "lazy fail" in some other use cases (such as intercepted transaction factory). Will try to find out a good solution to this problem. Long story, short, for now, will remove the checking in this create method.

Reply:
Left a TODO